sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -169,7 +174,10 @@ case class InMemoryTableScanExec(
   override def outputOrdering: Seq[SortOrder] =
     relation.child.outputOrdering.map(updateAttribute(_).asInstanceOf[SortOrder])
 
-  private def statsFor(a: Attribute) = relation.partitionStatistics.forAttribute(a)
+  // When we make canonicalized plan, we can't find a normalized attribute in this map.
+  // We return a `ColumnStatisticsSchema` for normalized attribute in this case.
```

Comment:
It looks weird to call `statsFor` on a canonicalized `InMemoryTableScanExec`, can we just make `buildFilter` a lazy val?

Reply:
I've tried that at beginning. However, `partitionFilters` uses `buildFilter`. Making `partitionFilters` a lazy doesn't work because when do `copy`, the initialization of `InMemoryTableScanExec` will try to materialize `partitionFilters` for coping it value.

Making `partitionFilters`, `buildFilter` as methods is not enough too, we also need to remove `@transient` from `relation` and `InMemoryRelation.partitionStatistics`. So I think it isn't worth and leave it as is.

Reply:
I don't get it. Regardless how copy is implemented in scala, ideally we can just mark `buildFilter` and `partitionFilters` as lazy, and in `doCanonicalize`, create a new `InMemoryTableScanExec`, which won't materialize `partitionFilters` in either the current `InMemoryTableScanExec` or the new `InMemoryTableScanExec`.

One problem I can think of is to serialize a canonicalized `InMemoryTableScanExec`, but it should never happen.

Reply:
Ah, sorry, I get it wrongly. The reason why it doesn't work is because `relation` is `@transient`. `partitionFilters` needs to be non-lazy, otherwise when we need to access `relation` in executor, we will get a `NullPointerException`.

Reply:
And I think it isn't worth removing @transient from `relation` and `InMemoryRelation.partitionStatistics` just for this. So I leave it as is.

Reply:
This can be solved if we add a `val stats = relation.partitionStatistics`, isn't it?

Reply:
Yes. I think so.